File: backend/app/utils/printer_models.py
"""Printer model normalization utilities.

Converts 3MF printer model names (e.g., "Bambu Lab X1 Carbon") to
normalized short names (e.g., "X1C") that match database storage.
"""

# Map from 3MF printer_model strings to normalized short names
PRINTER_MODEL_MAP = {
    "Bambu Lab X1 Carbon": "X1C",
    "Bambu Lab X1": "X1",
    "Bambu Lab X1E": "X1E",
    "Bambu Lab P1S": "P1S",
    "Bambu Lab P1P": "P1P",
    "Bambu Lab P2S": "P2S",
    "Bambu Lab A1": "A1",
    "Bambu Lab A1 Mini": "A1 Mini",
    "Bambu Lab A1 mini": "A1 Mini",
    "Bambu Lab A2L": "A2L",
    "Bambu Lab H2D": "H2D",
    "Bambu Lab H2D Pro": "H2D Pro",
    "Bambu Lab H2C": "H2C",
    "Bambu Lab H2S": "H2S",
    "Bambu Lab X2D": "X2D",
}

# Map from printer_model_id (internal codes in slice_info.config) to short names
# These are the codes Bambu Studio uses internally
PRINTER_MODEL_ID_MAP = {
    # X1 series
    "C11": "X1C",
    "C12": "X1",
    "C13": "X1E",
    # P1 series
    "P1P": "P1P",
    "P1S": "P1S",
    # P2 series
    "P2S": "P2S",
    # X2 series
    "N6": "X2D",
    # A1 series
    "A11": "A1",
    "A12": "A1 Mini",
    "N1": "A1 Mini",
    "N2S": "A1",
    "A04": "A1 Mini",
    # A2 series (A2L is single-FDM + integrated cutter/plotter — single nozzle)
    "N9": "A2L",
    # H2 series (Office/H series)
    "O1D": "H2D",
    "O1E": "H2D Pro",  # Some devices report O1E
    "O2D": "H2D Pro",  # Some devices report O2D
    "O1C": "H2C",
    "O1C2": "H2C",
    "O1S": "H2S",
}


# Rod/rail type classification for maintenance tasks.
# Carbon rods: X1, P1 series (CoreXY with carbon fiber rods)
# Steel rods: P2S, X2D series (hardened steel linear shafts)
# Linear rails: A1, A2, H2 series (linear rail motion system)
# Values must be uppercase with spaces stripped for normalized comparison.
CARBON_ROD_MODELS = frozenset(
    [
        # Display names (uppercase, no spaces)
        "X1",
        "X1C",
        "X1E",
        "P1P",
        "P1S",
        # Internal codes
        "C11",  # X1C
        "C12",  # X1
        "C13",  # X1E
    ]
)

STEEL_ROD_MODELS = frozenset(
    [
        # Display names (uppercase, no spaces)
        "P2S",
        "X2D",
        # Internal codes
        "N7",  # P2S
        "N6",  # X2D
    ]
)

LINEAR_RAIL_MODELS = frozenset(
    [
        # Display names (uppercase, no spaces)
        "A1",
        "A1MINI",
        "A2L",
        "H2D",
        "H2DPRO",
        "H2C",
        "H2S",
        # Internal codes
        "N1",  # A1 Mini
        "N2S",  # A1
        "N9",  # A2L
        "A04",  # A1 Mini (alternate)
        "A11",  # A1
        "A12",  # A1 Mini
        "O1D",  # H2D
        "O1E",  # H2D Pro
        "O2D",  # H2D Pro (alternate)
        "O1C",  # H2C
        "O1C2",  # H2C (dual nozzle variant)
        "O1S",  # H2S
    ]
)


# Models with an ethernet port.
# X1, P1P, A1, A1 Mini do NOT have ethernet.
ETHERNET_MODELS = frozenset(
    [
        # Display names (uppercase, no spaces)
        "X1C",
        "X1E",
        "X2D",
        "P1S",
        "P2S",
        "H2D",
        "H2DPRO",
        "H2C",
        "H2S",
        # Internal codes
        "C11",  # X1C
        "C13",  # X1E
        "N6",  # X2D
        "P1S",  # P1S
        "O1D",  # H2D
        "O1E",  # H2D Pro
        "O2D",  # H2D Pro (alternate)
        "O1C",  # H2C
        "O1C2",  # H2C (dual nozzle variant)
        "O1S",  # H2S
    ]
)


# Dual-nozzle (dual-extruder) printers. Single source of truth for nozzle
# class — consumed by ``BambuMQTTClient.start_print``, the K-profile routes,
# and the re-slice nozzle-class guard (previously an inline model tuple
# duplicated across all three). Re-slicing a model laid out for a single-nozzle
# printer onto one of these — or vice versa — is not yet supported: the source
# 3MF's embedded single-nozzle filament/extruder layout is not a valid
# dual-nozzle project and BambuStudio's multi-extruder validator rejects it.
DUAL_NOZZLE_MODELS = frozenset(
    [
        # Display names (uppercase, no spaces)
        "H2D",
        "H2DPRO",
        "H2C",
        "X2D",
        # Internal codes
        "O1D",  # H2D
        "O1E",  # H2D Pro
        "O2D",  # H2D Pro (alternate)
        "O1C",  # H2C
        "O1C2",  # H2C (dual nozzle variant)
        "N6",  # X2D
    ]
)


def has_ethernet(model: str | None) -> bool:
    """Return True if the printer model has an ethernet port."""
    if not model:
        return False
    normalized = model.strip().upper().replace(" ", "").replace("-", "")
    return normalized in ETHERNET_MODELS


def is_dual_nozzle_model(model: str | None) -> bool:
    """Return True if the printer model has two nozzles (H2D family / X2D)."""
    if not model:
        return False
    normalized = model.strip().upper().replace(" ", "").replace("-", "")
    return normalized in DUAL_NOZZLE_MODELS


def get_rod_type(model: str | None) -> str | None:
    """Return the rod/rail type for a printer model.

    Returns:
        "carbon" for X1/P1 series (carbon fiber rods),
        "steel_rod" for P2S/X2D series (hardened steel rods),
        "linear_rail" for A1/H2 series (linear rails),
        None for unknown models.
    """
    if not model:
        return None
    normalized = model.strip().upper().replace(" ", "").replace("-", "")
    if normalized in CARBON_ROD_MODELS:
        return "carbon"
    if normalized in STEEL_ROD_MODELS:
        return "steel_rod"
    if normalized in LINEAR_RAIL_MODELS:
        return "linear_rail"
    return None


def normalize_printer_model_id(model_id: str | None) -> str | None:
    """Convert printer_model_id (internal code) to normalized short name.

    Args:
        model_id: The printer_model_id from slice_info.config (e.g., "C11", "O1D")

    Returns:
        Normalized short name (e.g., "X1C", "H2D") or the original ID if unknown.
    """
    if not model_id:
        return None

    # Check known mappings
    if model_id in PRINTER_MODEL_ID_MAP:
        return PRINTER_MODEL_ID_MAP[model_id]

    # Return original if unknown (might already be a short name)
    return model_id


def normalize_printer_model(raw_model: str | None) -> str | None:
    """Convert 3MF printer_model to normalized short name.

    Args:
        raw_model: The printer_model string from 3MF metadata
            (e.g., "Bambu Lab X1 Carbon")

    Returns:
        Normalized short name (e.g., "X1C") or None if input is empty.
        Unknown models have "Bambu Lab " prefix stripped.
    """
    if not raw_model:
        return None

    # Check known mappings first
    if raw_model in PRINTER_MODEL_MAP:
        return PRINTER_MODEL_MAP[raw_model]

    # Strip "Bambu Lab " prefix for unknown models
    stripped = raw_model.replace("Bambu Lab ", "").strip()
    return stripped or None

File: frontend/src/utils/printerModels.ts
export type PrinterModelOptionGroup = {
  label: string;
  options: Array<{ value: string; label: string }>;
};

export const PRINTER_MODEL_GROUPS: PrinterModelOptionGroup[] = [
  {
    label: 'Bambu Lab A1 Series',
    options: [
      { value: 'A1', label: 'A1' },
      { value: 'A1 F', label: 'A1 F' },
      { value: 'A1 Mini', label: 'A1 Mini' },
    ],
  },
  {
    label: 'Bambu Lab A2 Series',
    options: [
      { value: 'A2L', label: 'A2L' },
    ],
  },
  {
    label: 'Bambu Lab P Series',
    options: [
      { value: 'P1P', label: 'P1P' },
      { value: 'P1S', label: 'P1S' },
      { value: 'P2S', label: 'P2S' },
    ],
  },
  {
    label: 'Bambu Lab X Series',
    options: [
      { value: 'X1', label: 'X1' },
      { value: 'X1C', label: 'X1 Carbon' },
      { value: 'X1E', label: 'X1E' },
      { value: 'X2D', label: 'X2D' },
    ],
  },
  {
    label: 'Bambu Lab H Series',
    options: [
      { value: 'O1C', label: 'O1C' },
      { value: 'O1E', label: 'O1E' },
      { value: 'O1S', label: 'O1S' },
      { value: 'H2C', label: 'H2C' },
      { value: 'H2D', label: 'H2D' },
      { value: 'H2D Pro', label: 'H2D Pro' },
      { value: 'H2S', label: 'H2S' },
    ],
  },
  {
    label: 'Elegoo',
    options: [
      { value: 'Elegoo Neptune 3', label: 'Neptune 3' },
      { value: 'Elegoo Neptune 3 Pro', label: 'Neptune 3 Pro' },
      { value: 'Elegoo Neptune 3 Plus', label: 'Neptune 3 Plus' },
      { value: 'Elegoo Neptune 3 Max', label: 'Neptune 3 Max' },
      { value: 'Elegoo Neptune 4', label: 'Neptune 4' },
      { value: 'Elegoo Neptune 4 Pro', label: 'Neptune 4 Pro' },
      { value: 'Elegoo Neptune 4 Plus', label: 'Neptune 4 Plus' },
      { value: 'Elegoo Neptune 4 Max', label: 'Neptune 4 Max' },
      { value: 'Elegoo Centauri', label: 'Centauri' },
      { value: 'Elegoo Centauri Carbon', label: 'Centauri Carbon' },
    ],
  },
  {
    label: 'Voron',
    options: [
      { value: 'Voron V0.2', label: 'V0.2' },
      { value: 'Voron Trident', label: 'Trident' },
      { value: 'Voron 2.4', label: '2.4' },
      { value: 'Voron Switchwire', label: 'Switchwire' },
      { value: 'Voron Legacy', label: 'Legacy' },
    ],
  },
  {
    label: 'Creality Klipper',
    options: [
      { value: 'Creality Ender-3', label: 'Ender-3' },
      { value: 'Creality Ender-3 Pro', label: 'Ender-3 Pro' },
      { value: 'Creality Ender-3 V2', label: 'Ender-3 V2' },
      { value: 'Creality Ender-3 S1', label: 'Ender-3 S1' },
      { value: 'Creality Ender-5 Plus', label: 'Ender-5 Plus' },
      { value: 'Creality CR-10S Pro', label: 'CR-10S Pro' },
      { value: 'Creality CR-10S Pro V2', label: 'CR-10S Pro V2' },
      { value: 'Creality K1', label: 'K1' },
      { value: 'Creality K1C', label: 'K1C' },
      { value: 'Creality K2', label: 'K2' },
      { value: 'Creality K2 Pro', label: 'K2 Pro' },
      { value: 'Creality K2 Plus', label: 'K2 Plus' },
    ],
  },
  {
    label: 'Snapmaker Klipper',
    options: [
      { value: 'Snapmaker U1', label: 'U1' },
    ],
  },
  {
    label: 'Prusa',
    options: [
      { value: 'Prusa CORE One', label: 'CORE One' },
      { value: 'Prusa MK4S', label: 'MK4S' },
      { value: 'Prusa MK4', label: 'MK4' },
      { value: 'Prusa MK3.9S', label: 'MK3.9S' },
      { value: 'Prusa MK3.9', label: 'MK3.9' },
      { value: 'Prusa MK3.5S', label: 'MK3.5S' },
      { value: 'Prusa MK3.5', label: 'MK3.5' },
      { value: 'Prusa XL', label: 'XL' },
      { value: 'Prusa MINI+', label: 'MINI+' },
      { value: 'Prusa MK3S+', label: 'MK3S+' },
      { value: 'Prusa SL1S SPEED', label: 'SL1S SPEED' },
    ],
  },
  {
    label: 'Generic',
    options: [
      { value: 'Klipper', label: 'Klipper / Moonraker' },
      { value: 'PrusaLink', label: 'PrusaLink' },
      { value: 'Generic Klipper Printer', label: 'Generic Klipper Printer' },
      { value: 'Generic FDM Printer', label: 'Generic FDM Printer' },
    ],
  },
];

// Map SSDP model codes to display names
export function mapModelCode(ssdpModel: string | null): string {
  if (!ssdpModel) return '';
  const modelMap: Record<string, string> = {
    // H2 Series
    O1D: 'H2D',
    O1E: 'H2D Pro',
    O2D: 'H2D Pro',
    O1C: 'H2C',
    O1C2: 'H2C',
    O1S: 'H2S',
    // X1 Series
    'BL-P001': 'X1C',
    'BL-P002': 'X1',
    'BL-P003': 'X1E',
    // X2 Series
    N6: 'X2D',
    // P Series
    C11: 'P1S',
    C12: 'P1P',
    C13: 'P2S',
    // A1 Series
    N2S: 'A1',
    N1: 'A1 Mini',
    // A2 Series
    N9: 'A2L',
    // Direct matches
    X1C: 'X1C',
    X1: 'X1',
    X1E: 'X1E',
    X2D: 'X2D',
    P1S: 'P1S',
    P1P: 'P1P',
    P2S: 'P2S',
    A1: 'A1',
    'A1 Mini': 'A1 Mini',
    A2L: 'A2L',
    H2D: 'H2D',
    'H2D Pro': 'H2D Pro',
    H2C: 'H2C',
    H2S: 'H2S',
  };
  return modelMap[ssdpModel] || ssdpModel;
}
